### ISSUE TITLE
Suppress announcements for highway name changes

### ISSUE THREAD

**Issue by danpat** [Project-OSRM/osrm-backend#3281]
Routing across these two ways:

http://www.openstreetmap.org/way/27652088
http://www.openstreetmap.org/way/48570003

Leads to a "Continue straight" announcement.

This seems superfluous - you're on a one-way highway with nowhere else to go.  Announcing the name change seems like a distraction.

@MoKob should we be suppressing these situations?

http://map.project-osrm.org/?z=15&center=37.739770%2C-122.417607&loc=37.757144%2C-122.392588&loc=37.740818%2C-122.398885&hl=en&alt=0

![screen shot 2016-11-10 at 12 59 54 pm](https://cloud.githubusercontent.com/assets/1892250/20194671/340bbe84-a748-11e6-849c-31798aad6a93.png)

**Comment by 1ec5** [comment]
Previously #2744.

**Comment by MoKob** [comment]
We had this discussion a few times. Right now I don't have any other tool than announcing a name to give the visual clue that the highway has changed. It might come in necessary later, since you might actually want to follow names that show a different name.

Consider the following scenario: You switch from `A` to `B` in such a location.
Not soon after, you see highway signs and they only show `B`. Now you are looking for `A` but can't find it anywhere. Not getting any instruction (since staying on `B` is considered obvious here), you get confused.

In my view we now have a few choices:

1) Decide not to announce the instruction but use it as pure visual update in the `frontend` (possible, since you get the information from the intersections field: only two entries, one of them allowed).

2) completely omit the information in the `backend` for all undisturbed name changes - Unknown implications for later turns (see above), especially as we can see this a lot within cities as well. Here we would be on a different street than expected, which can (especially in rural Germany) be very confusing.

3) only omit the information on motorways: still not as safe, since motorways are often tagged within cities as well and the above sign scenario still holds true (don't just think names, also references)

4) introduce a new instruction type that directly marks these kinds as name update, but I would feel the `visual name update` and `new name` next to each other would just be too confusing.

**Comment by miccolis** [comment]
I want to flag this for @mcwhittemore as he's thinking about these sorts of problems.

I'd be interested to hear from both @MoKob and @mcwhittemore as to whether we can delete this decision to downstream consumers and not make a change in OSRM.

**Comment by mcwhittemore** [comment]
I'm not a bit fan that the name change creates a maneuver as this is less about what the car is doing and more about what the road is doing, but currently maneuvers are the only way for OSRM to pass along name changes and I know of at least one highly used diving app that shows the current road name on the screen at all times.

For now, this should be punted to down stream post processors to think about reformatting. Maybe some of them will come up with a good idea and can present back a cleaner api for this kind of information to OSRM.

**Comment by MoKob** [comment]
@miccolis we already have quite some overhead in logic surrounding the suppression of name changes. For example bridges, tunnels, short segments in segregated intersections and even just in general. A maneuver, as defined in our terms, is something where we require action from a driver.
The name change, as is in there, currently and always was a workaround on how to pass information to a driver about the current road.
My feeling here is that we shouldn't favour keeping a, already kind of messed up, workaround, for something that can be handled in a clear way.
To me, passing names along the route should be a thing for the route-annotator or some OSRM internal annotation feature. Leaving it in the API leaves us with a mix of general hints and actual maneuver.

Situation right now:

 - OSRM uses a heuristic to distinguish between ref/name combinations to decide what is the same/not
 - new names get announced if they remain valid for long
 - new names get suppressed if short / change back to previous name shortly after

With https://github.com/Project-OSRM/osrm-backend/pull/4398 I've cleared up the current mix of announcing some but not all name changes. I am fine with having this in the pipeline for a bit, but ultimately I would strongly suggest to rip out name-changes from OSRM to emit only clear maneuvers in the steps.

**Comment by miccolis** [comment]
> ultimately I would strongly suggest to rip out name-changes from OSRM to emit only clear maneuvers in the steps

I'm in total agreement and it sounds like we've got broad consensus about this.

What I think we need to figure out is; in what circumstances we'd want to maintain some name/ref changes, how we'd want to represent that in the future and what we need short term. @mcwhittemore of the options @MoKob [posted above](https://github.com/Project-OSRM/osrm-backend/issues/3281#issuecomment-259913820) what would you lean towards, or are you seeing something else?

**Comment by mcwhittemore** [comment]
> Decide not to announce the instruction but use it as pure visual update in the frontend

This seems like the cleanest option. We'd want to do some work to expand the [intersection object](http://project-osrm.org/docs/v5.10.0/api/#intersection-object) to include  bit more information such as the name, refs and destinations of the road the user will be on after the intersection, but generally speaking I think this option should work.

Here is how I'd add the needed info to the intersection object? @MoKob is there a better way to pass this info along? Is there a way to infer this without passing it along?

```json
{
    "location":[13.394718,52.543096],
    "in":0,
    "out":2,
    "bearings":[60,150,240,330],
    "entry":["false","true","true","true"],
    "classes": ["toll", "restricted"],
    "lanes":{
        "indications": ["left", "straight"],
        "valid": "false"
    },
    "roadAfterIntersection": {
        "name": "some name",
        "ref": "some refs",
        "destination": "some destinations"
    },
}
```

**Comment by MoKob** [comment]
@mcwhittemore I agree that the intersection would probably be the best location to put that information. With the way you propose I would have a few questions / concerns:

### Location/Naming

roadAfterIntersection is essentially what the intersection array indicates. We have information associated with the turn itself (which is a bit confusing, since lanes are actually on the segment prior to the turn) and information about where the turn is leading us to.
Using `afterIntersection` is essentially just iterating on the concept.
Is there a reason not to directly use `name/ref/destination` as field names? If so, how about using a more neutral name like `signage` or `cues`?

### Response size

If we put the name/ref/destination into every item, we add a lot of repeated information. I don't think that the issue is a real issue, but we should be aware of it. Possible solutions here would be to only post updates (whenever we have the data actually change), or to use a lookup table and index the names. I would prefer to keep it simple, though, and simply eat the cost in response size.

### Deprecate in step?

With the newly added data, we would at least duplicate the information currently present in the RouteStep. I would suggest, that we deprecate that information in the next API version, if we should go with the annotations in the intersections array.

**Comment by mcwhittemore** [comment]
> roadAfterIntersection is essentially what the intersection array indicates

It would be nice if `{name, ref, destination}` was an object on the intersection object. `signage` sounds good to me. `cues` seems like it could mean more than what we're talking about right now or even something else so I think we should save that word for later.

> If we put the name/ref/destination into every item, we add a lot of repeated information

I think it should be fine to do this for now. It might be worth doing a quick check to make sure this doesn't increase the response size by 2x, but I doubt it will.

> I would suggest, that we deprecate that information in the next API version

Agreed. Indicating that we know something will change ASAP is a good idea. That said, this won't be dropped until OSRM@6.0.0 right?

**Comment by 1ec5** [comment]
> It would be nice if `{name, ref, destination}` was an object on the intersection object. `signage` sounds good to me. `cues` seems like it could mean more than what we're talking about right now or even something else so I think we should save that word for later.

We could solve #3329 at the same time by giving each intersection object `name`, `ref`, `destination`, `destination_ref`, and `pronunciation` properties set to arrays instead of strings that are ambiguous and difficult to parse.

**Comment by mcwhittemore** [comment]
Thinking about this a bit more, we should add `names`, `refs` and `destinations` to the Intersection object as arrays. This would align with `entry` and `bearings` and thus fit into an established paradigm.

**Comment by 1ec5** [comment]
> we should add `names`, `refs` and `destinations` to the Intersection object as arrays

This refers to the names/refs/destinations of the road on which the step travels, right? Not the names/refs/destinations of cross streets, as proposed in #4379?
